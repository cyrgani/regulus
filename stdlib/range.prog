# Returns a list of all integers from `start` to `end`, including `start` but not `end`.
# Rust equivalent: `start..end`
def(range, start, end, _(
    if(>(start, end), error("cannot construct range with start > end")),
    =(i, start),
    =(l, list()),
    while(<(i, end), _(
        =(l, push(l, i)),
        =(i, +(i, 1)),
    )),
    l
)),

=(.., range)