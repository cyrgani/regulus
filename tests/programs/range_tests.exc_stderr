RangeError: cannot construct range with start > end
at <stl:range>:4:28
at <stl:range>:4:7
at <stl:range>:3:25
at programs/range_tests.re:6:37
at programs/range_tests.re:6:31

RangeError: cannot construct range with start > end
at <stl:range>:4:28
at <stl:range>:4:7
at <stl:range>:3:25
at programs/range_tests.re:7:37
at programs/range_tests.re:7:31

