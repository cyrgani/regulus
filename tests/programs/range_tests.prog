import("range"),
assert_eq(range(0, 4), list(0, 1, 2, 3)),
assert_eq(range(0, 1), list(0)),
assert_eq(range(-2, 2), list(-2, -1, 0, 1)),
assert_eq(range(0, 0), list()),
assert_eq(catch(range(1, 0)), "UserRaised: cannot construct range with start > end"),
assert_eq(catch(range(1, -1)), "UserRaised: cannot construct range with start > end"),

assert_eq(..(0, 4), list(0, 1, 2, 3)),
