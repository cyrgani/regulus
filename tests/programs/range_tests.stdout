UserRaisedError: cannot construct range with start > end 
UserRaisedError: cannot construct range with start > end 
